ext/winevt/winevt_subscribe.c

```diff
@@ -144,17 +153,61 @@ rb_winevt_subscribe_subscribe(int argc, VALUE* argv, VALUE self)
   return Qfalse;
 }
 
+BOOL
+rate_limit_check_handler(struct WinevtSubscribe *winevtSubscribe)
+{
+  time_t now;
+  ULONG currRate = 0;
+
+  if (winevtSubscribe->rateLimit == SUBSCRIBE_RATE_INFINITE)
+    return FALSE;
+
+  time(&now);
+
+  winevtSubscribe->previousRate = 0;
+
+  if (now <= winevtSubscribe->lastTime) {
+    winevtSubscribe->previousRate = winevtSubscribe->currentRate;
+    currRate = winevtSubscribe->currentRate;
+    if (currRate >= winevtSubscribe->rateLimit) {
+      return TRUE;
+    }
+  }
+
+  return FALSE;
+}
+
+void
+rate_limit_state_handler(struct WinevtSubscribe *winevtSubscribe, ULONG count)
+{
+  time_t lastTime = 0;
+  ULONG currRate = 0;
+
+  if (winevtSubscribe->rateLimit == SUBSCRIBE_RATE_INFINITE)
+    return;
+
+  time(&lastTime);
+  winevtSubscribe->lastTime = lastTime;
+  currRate = winevtSubscribe->previousRate;
+  currRate += count;
+  winevtSubscribe->currentRate = currRate;
+}
+
 static VALUE
 rb_winevt_subscribe_next(VALUE self)
 {
   EVT_HANDLE hEvents[SUBSCRIBE_ARRAY_SIZE];
-  ULONG count;
+  ULONG count = 0;
   DWORD status = ERROR_SUCCESS;
   struct WinevtSubscribe* winevtSubscribe;
 
   TypedData_Get_Struct(
     self, struct WinevtSubscribe, &rb_winevt_subscribe_type, winevtSubscribe);
 
+  if (rate_limit_check_handler(winevtSubscribe)) {
```

Comment:
It would be better that the function names for `rate_limit_check_handler` and `rate_limit_state_handler` are more readable.

* `handler` isn't needed any more because they are called directly instead of vi function pointers.
* `rate_limit_check_handler` should clarify the meaning of its return value (e.g. `is_rate_limited`)

Reply:
Fixed in https://github.com/fluent-plugins-nursery/winevt_c/pull/9/commits/a6fb62d618af245940aef6e6db56e64806426896.